dev/tasks/crossbow.py

```diff
@@ -391,45 +422,101 @@ def submit(task_names, job_prefix, config_path, dry_run, arrow_path,
+            click.echo(filename + click.style(statemsg, fg=COLORS[state]))
 
 
 @crossbow.command()
 @click.argument('job-name', required=True)
-@click.option('--target-dir', default=DEFAULT_ARROW_PATH,
+@click.option('--upload/--no-upload', default=True,
+              help='Upload the signatures to github releases')
+@click.option('--gpg-homedir', default=None,
+              help=('Full pathname to directory containing the public and '
+                    'private keyrings. Default is whatever GnuPG defaults to'))
+@click.option('--target-dir', default=DEFAULT_ARROW_PATH / 'pkgs',
               help='Directory to download the build artifacts')
 @click.option('--queue-path', default=DEFAULT_QUEUE_PATH,
               help='The repository path used for scheduling the tasks. '
                    'Defaults to crossbow directory placed next to arrow')
 @github_token
-def artifacts(job_name, target_dir, queue_path, github_token):
-    queue = Queue(queue_path)
-    username, reponame = queue.parse_user_repo()
+def sign(job_name, upload, gpg_homedir, target_dir, queue_path, github_token):
+    """Download and sign build artifacts from github releases"""
 
-    gh = github3.login(token=github_token)
-    repo = gh.repository(username, reponame)
-    release = repo.release_from_tag(job_name)
+    import gnupg
+    gpg = gnupg.GPG(gnupghome=gpg_homedir)
+
+    # initialize and fetch the queue repository
+    queue = Queue(queue_path)
+    queue.fetch()
 
-    for asset in release.assets():
-        click.echo('Downloading asset {} ...'.format(asset.name))
-        asset.download(target_dir / asset.name)
+    # query the job's artifacts
+    job = queue.get(job_name)
+    assets = queue.github_assets(job, token=github_token)
+
+    click.echo('Downloading and signing assets...')
+    target_dir = Path(target_dir).absolute()
+    target_dir.mkdir(exist_ok=True)
+
+    tpl = '[{:>8}] '
+    signatures = []
+    for task_name, task in job.tasks.items():
+        for artifact in task.artifacts:
+            if artifact not in assets:
+                msg = click.style(tpl.format('MISSING'), fg=COLORS['missing'])
+                click.echo(msg + fname)
+                continue
+
+            # download artifact
+            target_path = target_dir / artifact
+            assets[artifact].download(target_path)
+
+            # sign the artifact
+            with target_path.open('rb') as fp:
+                signature_path = Path(str(target_path) + '.sig')
+                gpg.sign_file(fp, detach=True, clearsign=False,
+                              output=str(signature_path))
+
+            msg = click.style(tpl.format('SIGNED'), fg=COLORS['ok'])
+            click.echo(msg + artifact)
+
+            signatures.append(signature_path)
+
+    # upload the signatures
+    if upload:
+        queue.upload_assets(job, signatures, token=github_token,
+                            content_type='application/pgp-signature')
```

Comment:
What's the thinking re: uploading the GPG signatures? We only need the sigs right before uploading the artifacts to the ASF's dist system right before a release vote

Reply:
I'm not exactly aware of the steps, I thought the release vote uses the assets uploaded to github, then comes the ASF's dist system (which is what?), bintray, forge, etc.

Reply:
BTW here are the artifacts and signatures: https://github.com/kszucs/crossbow/releases

Should I skip uploading the signatures?

Reply:
Uploading signatures in the script is not necessary. A release manager should compute the signatures on all artifacts locally before they are committed to the ASF svn for distribution.